memiavl/db.go

```diff
@@ -302,9 +342,68 @@ func (db *DB) RewriteSnapshotBackground() error {
 }
 
 func (db *DB) Close() error {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
 	return errors.Join(db.MultiTree.Close(), db.wal.Close())
 }
 
+// TreeByName wraps MultiTree.TreeByName to add a lock.
+func (db *DB) TreeByName(name string) *Tree {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	return db.MultiTree.TreeByName(name)
+}
+
+// Hash wraps MultiTree.Hash to add a lock.
+func (db *DB) Hash() []byte {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	return db.MultiTree.Hash()
+}
+
+// Version wraps MultiTree.Version to add a lock.
+func (db *DB) Version() int64 {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	return db.MultiTree.Version()
+}
+
+// LastCommitInfo returns the last commit info.
+func (db *DB) LastCommitInfo() *storetypes.CommitInfo {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	return db.MultiTree.LastCommitInfo()
+}
```

Comment:
do we need make a lock wrapper like?
```suggestion
func withLock[T any](db *DB, f func(*MultiTree) T) T {
	db.mtx.Lock()
	defer db.mtx.Unlock()
	return f(&db.MultiTree)
}

// LastCommitInfo wraps MultiTree LastCommitInfo to add a lock.
func (db *DB) LastCommitInfo() *storetypes.CommitInfo {
	return withLock(db, func(mtree *MultiTree) *storetypes.CommitInfo {
		return mtree.LastCommitInfo()
	})
}
```

Reply:
the non-callback version seems simple enough?

Reply:
yea